lib/streamlit/commands/query_params.py

```diff
@@ -80,7 +92,48 @@ def set_query_params(**query_params: Any) -> None:
     ctx = get_script_run_ctx()
     if ctx is None:
         return
-    ctx.query_string = parse.urlencode(query_params, doseq=True)
+
     msg = ForwardMsg()
-    msg.page_info_changed.query_string = ctx.query_string
+    msg.page_info_changed.query_string = _ensure_no_embed_params(
+        query_params, ctx.query_string
+    )
     ctx.enqueue(msg)
+
+
+def _ensure_no_embed_params(
+    query_params: Dict[str, List[str]], query_string: str
+) -> str:
+    """Ensures there are no embed params set (raises StreamlitAPIException) if there is a try,
+    also makes sure old param values in query_string are preserved. Returns query_string : str."""
+    # Get query params dict without embed, embed_options params
+    query_params_without_embed = util.exclude_key_query_params(
+        query_params, keys_to_exclude=EMBED_QUERY_PARAMS_KEYS
+    )
+    if query_params != query_params_without_embed:
+        raise StreamlitAPIException(
+            "Query param embed and embed_options (case-insensitive) cannot be set using set_query_params method."
+        )
+
+    all_current_params = parse.parse_qs(query_string)
+    current_embed_params = "".join(
+        [
+            f"&{EMBED_QUERY_PARAM}={param}"
+            for param in util.extract_key_query_params(
+                all_current_params, param_key=EMBED_QUERY_PARAM
+            )
+        ]
+    )
+    current_embed_options_params = "".join(
+        [
+            f"&{EMBED_OPTIONS_QUERY_PARAM}={param}"
+            for param in util.extract_key_query_params(
+                all_current_params, param_key=EMBED_OPTIONS_QUERY_PARAM
+            )
+        ]
+    )
+    current_embed_params += current_embed_options_params
```

Comment:
Actually this can be rewritten as:

```
current_embed_params = parse.urlencode(
    {
       EMBED_QUERY_PARAM: [param for param in util.extract_key_query_params(
                all_current_params, param_key=EMBED_QUERY_PARAM)],
       EMBED_OPTIONS_QUERY_PARAM: [param for param in util.extract_key_query_params(
                all_current_params, param_key=EMBED_OPTIONS_QUERY_PARAM)]
     },
     doseq=True)
```

Reply:
True this is much more elegant, thank you Michal 🙇